### PR TITLE
Added locking for build jobs

### PR DESCRIPTION
**Pull request by Luzifer**

### CODE REVIEW

**Comment by Luzifer** [comment]
@jbenet PTAL

File: cmd/starter/main.go [hunk 1]
```diff
@@ -130,6 +130,24 @@ func fetchBuildJob() {
...
+
+	if lock != nil {
+		// If build-job is currently locked, try again later
+		queueEntry, err := job.ToByte()
+		orFail(err)
+		redisClient.RPush("build-queue", string(queueEntry))
+		return
+	}
+
+	redisClient.Set(fmt.Sprintf("project::%s::build-lock", repo), "locked", 1800, 0, false, false)
```
Comment (jbenet) [inline cmd/starter/main.go]: i think there's a TOCTTOU bug here. what happens if:

```
goroutine1: calls redis get, and it completes. (lock == nil)
goroutine2: calls redis get, and it completes. (lock == nil)
goroutine2: calls redis set ("has lock")
goroutine1: calls redis set ("has lock")
```

I think you need one of the redis atomic ops? (not clear to me this handles it)

**Comment by Luzifer** [comment]
@jbenet This new solution should solve that problem, right?

File: cmd/starter/main.go [hunk 2]
```diff
@@ -130,6 +131,26 @@ func fetchBuildJob() {
...
+		return
+	}
+
+	if string(lock) != lockID {
+		// If build-job is currently locked but not by us, try again later
+		queueEntry, err := job.ToByte()
+		orFail(err)
+		redisClient.RPush("build-queue", string(queueEntry))
+		return
+	}
```
Comment (jbenet) [inline cmd/starter/main.go:152]: Nope, imagine:

```
proc1> redisClient.Set()
proc1> redisClient.Get()
ctx switch
proc2> redisClient.Set()
proc2> redisClient.Get()

now both proc1 and proc2 have lock == their own lockID.
```

Comment (jbenet) [inline cmd/starter/main.go:152]: TAL at https://github.com/hjr265/redsync.go

https://github.com/hjr265/redsync.go/blob/master/redsync/mutex.go

Comment (jbenet) [inline cmd/starter/main.go:152]: Ohh maybe, just caught the `mustNotExists=true` in the `Set` call. http://redis.io/commands/set  claims something sort of atomic. maybe it returns an error if it fails?

Comment (Luzifer) [inline cmd/starter/main.go:152]: I don't really know. The code for the set command is quite intransparent… Thats why I wrote the get check to ensure the own ID is set to the lock. Any other sets should fail and the id does not match so the get returns from the function…

Comment (jbenet) [inline cmd/starter/main.go:152]: should grab someone with redis knowledge. maybe ask in #ipfs ?

**Comment by Luzifer** [comment]
Heard different opinions, mostly "it might work but dunno". Will just merge it and see what happens.

**Pull request merged**